src/parcsr_ls/par_ilu_setup.c

```diff
@@ -407,311 +398,494 @@ hypre_ILUSetup( void               *ilu_vdata,
-                                  ilu_solve_policy, &ilu_solve_buffer,
-                                  &matBL_info, &matBU_info, &matSL_info, &matSU_info, &matBLU_d, &matS,
-                                  &matE_d, &matF_d, &A_diag_fake, tri_solve);//BJ + hypre_ilut(), setup the device solve
-#else
-         hypre_ILUSetupILUT(matA, max_row_elmts, droptol, perm, perm, n, n, &matL, &matD, &matU, &matS,
-                            &u_end); //BJ + hypre_ilut()
+         if (exec == HYPRE_EXEC_DEVICE)
+         {
+#if !defined(HYPRE_USING_UNIFIED_MEMORY)
+            hypre_error_w_msg(HYPRE_ERROR_GENERIC,
+                              "ILUT setup on device runs requires unified memory!");
+            return hypre_error_flag;
 #endif
+            /* BJ + hypre_ilut(), setup the device solve */
+            hypre_ILUSetupILUTDevice(matA, max_row_elmts, droptol, perm, perm,
+                                     n, n, matL_des, matU_des,
+                                     ilu_solve_policy, &ilu_solve_buffer,
+                                     &matBL_info, &matBU_info, &matSL_info,
+                                     &matSU_info, &matBLU_d, &matS, &matE_d,
+                                     &matF_d, &A_diag_fake, tri_solve);
+         }
+         else
+#endif
+         {
+            /* BJ + hypre_ilut() */
+            hypre_ILUSetupILUT(matA, max_row_elmts, droptol, perm, perm, n, n,
+                               &matL, &matD, &matU, &matS, &u_end);
+         }
          break;
+
       case 10:
 #if defined(HYPRE_USING_CUDA) && defined(HYPRE_USING_CUSPARSE)
-         if (fill_level == 0)
+         if (exec == HYPRE_EXEC_DEVICE)
          {
-            /* Only support ILU0 */
-            hypre_ILUSetupILU0Device(matA, perm, perm, n, nLU, matL_des, matU_des, ilu_solve_policy,
-                                     &ilu_solve_buffer,
-                                     &matBL_info, &matBU_info, &matSL_info, &matSU_info, &matBLU_d, &matS,
-                                     &matE_d, &matF_d, &A_diag_fake, 1);//BJ + cusparse_ilu0()
+            if (fill_level == 0)
+            {
+               /* BJ + cusparse_ilu0() - Only support ILU0 */
+               hypre_ILUSetupILU0Device(matA, perm, perm, n, nLU, matL_des, matU_des,
+                                        ilu_solve_policy, &ilu_solve_buffer,
+                                        &matBL_info, &matBU_info, &matSL_info,
+                                        &matSU_info, &matBLU_d, &matS, &matE_d,
+                                        &matF_d, &A_diag_fake, 1);
+            }
+            else
+            {
+#if !defined(HYPRE_USING_UNIFIED_MEMORY)
+               hypre_error_w_msg(HYPRE_ERROR_GENERIC,
+                                 "GMRES+ILUK setup on device runs requires unified memory!");
+               return hypre_error_flag;
+#endif
+
+               /* BJ + cusparse_ilu0() */
```

Comment:
Same comment as above. GMRES + ILUK

Reply:
Done in [7c78be2](https://github.com/hypre-space/hypre/pull/816/commits/7c78be2eb456fc781ea7af9e9e9e8a07f5681484)